mlflow/gateway/handlers.py

```diff
@@ -0,0 +1,225 @@
+from enum import Enum
+import json
+from pathlib import Path
+from pydantic import BaseModel, validator, parse_obj_as
+from pydantic.json import pydantic_encoder
+from typing import Optional, Union, List, Dict, Any
+import yaml
+
+from mlflow.exceptions import MlflowException
+from mlflow.gateway.constants import PROVIDERS
+from mlflow.gateway.utils import is_valid_endpoint_name, check_configuration_route_name_collisions
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+
+
+class Provider(str, Enum):
+    UNSPECIFIED_PROVIDER = "custom"
+    OPENAI = "openai"
+    ANTHROPIC = "anthropic"
+    DATABRICKS_SERVING_ENDPOINT = "databricks_serving_endpoint"
+    MLFLOW = "mlflow"
+
+
+class RouteType(str, Enum):
+    CUSTOM = "custom"
+    LLM_V1_COMPLETIONS = "llm/v1/completions"
+    LLM_V1_CHAT = "llm/v1/chat"
+
+
+class OpenAIConfig(BaseModel):
+    openai_api_key: Optional[str] = None
+    openai_api_key_env_var: Optional[str] = None
+    openai_api_type: Optional[str] = None
+    openai_api_base: Optional[str] = "https://api.openai.com/v1"
+    openai_api_version: Optional[str] = None
+    openai_organization: Optional[str] = None
+
+
+class AnthropicConfig(BaseModel):
+    anthropic_api_key: Optional[str] = None
+    anthropic_api_key_env_var: Optional[str] = None
+    anthropic_api_base: Optional[str] = "https://api.anthropic.com/"
+
+
+class DatabricksConfig(BaseModel):
+    databricks_api_token: Optional[str] = None
+    databricks_api_token_env_var: Optional[str] = None
+    databricks_api_base: str
+
+
+class MLflowConfig(BaseModel):
+    api_base: str
+
+
+class CustomConfig(BaseModel):
+    api_key: Optional[str] = None
+    api_key_env_var: Optional[str] = None
+    api_base: str
+    api_version: Optional[str] = None
+
+
+config_types = {
+    Provider.OPENAI: OpenAIConfig,
+    Provider.ANTHROPIC: AnthropicConfig,
+    Provider.DATABRICKS_SERVING_ENDPOINT: DatabricksConfig,
+    Provider.MLFLOW: MLflowConfig,
+    Provider.UNSPECIFIED_PROVIDER: CustomConfig,
+}
+
+
+class ModelInfo(BaseModel):
+    name: Optional[str] = None
+    provider: Provider = Provider.UNSPECIFIED_PROVIDER
+
+
+def _config_instance_validation(config, provider):
```

Comment:
Does pydantic take care of this?

Reply:
Pydantic doesn't have the ability to handle the optional (one of these must be set) logic for how we're handling the requirement of either an api key or an environment config pointing to an api key. 
For the other required validations that are present, by not setting them as Optional, we get free structural validation from pydantic. It's just this configuration aspect that we need to handle the logic outside of the typical `@validator` logic.

Reply:
The presence of the base_route config validation is purely to give a more clear exception, though. The default behavior in validation won't work like other validation steps since we have to parse the `config` data as a Dict so that we can do the key validation logic (if we directly apply the config classes, we miss out on the ability to do the validation for 'at least one of these must be present'